Sources/NIO/Codec.swift

```diff
@@ -790,3 +790,266 @@ extension MessageToByteHandler {
+///                 }
+///             } catch {
+///                 context.fireErrorCaught(error)
+///             }
+///         }
+///
+///         // Forward on the body messages as whole messages
+///         func channelReadMessage(context: ChannelHandlerContext, message: String) {
+///             context.fireChannelRead(self.wrapInboundOut(.body(message)))
+///         }
+///     }
+///
+///     private class DecodedBodyHTTPHandler: ChannelInboundHandler {
+///         typealias InboundIn = HTTPPart<HTTPRequestHead, String>
+///         typealias OutboundOut = HTTPServerResponsePart
+///
+///         var msgs: [String] = []
+///
+///         func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+///             let message = self.unwrapInboundIn(data)
+///
+///             switch message {
+///             case .head(let head):
+///                 print("head: \(head)")
+///             case .body(let msg):
+///                 self.msgs.append(msg)
+///             case .end(let trailers):
+///                 print("trailers: \(trailers)")
+///                 var responseBuffer = context.channel.allocator.buffer(capacity: 32)
+///                 for msg in msgs {
+///                     responseBuffer.writeString(msg)
+///                     responseBuffer.writeStaticString("\n")
+///                 }
+///                 var headers = HTTPHeaders()
+///                 headers.add(name: "content-length", value: String(responseBuffer.readableBytes))
+///
+///                 context.write(self.wrapOutboundOut(HTTPServerResponsePart.head(
+///                     HTTPResponseHead(version: .init(major: 1, minor: 1),
+///                                      status: .ok, headers: headers))), promise: nil)
+///
+///                 context.write(self.wrapOutboundOut(HTTPServerResponsePart.body(
+///                     .byteBuffer(responseBuffer))), promise: nil)
+///                 _ = context.writeAndFlush(self.wrapOutboundOut(HTTPServerResponsePart.end(nil)))
+///             }
+///         }
+///     }
+///
+///     let group = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+///     let bootstrap = ServerBootstrap(group: group).childChannelInitializer({channel in
+///         channel.pipeline.configureHTTPServerPipeline(withPipeliningAssistance: true, withErrorHandling: true).flatMap { _ in
+///             channel.pipeline.addHandlers([RawBodyMessageHandler(), DecodedBodyHTTPHandler()])
+///         }
+///     })
+///     let channelFuture = bootstrap.bind(host: "127.0.0.1", port: 0)
+///
+public class NIOSingleStepByteToMessageProcessor<Decoder: NIOSingleStepByteToMessageDecoder> {
+    private enum DecodeMode {
+        /// This is a usual decode, ie. not the last chunk
+        case normal
+        /// Last chunk
+        case last
+    }
+
+    private var decoder: Decoder
+    private let maximumBufferSize: Int?
+    internal private(set) var buffer: ByteBuffer?
```

Comment:
Is this only `internal` for testing purposes? If so then please prefix `buffer` with an underscore.